adapters/algorix/algorix.go

```diff
@@ -0,0 +1,163 @@
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("x-openrtb-version", "2.5")
+
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     endPoint,
+		Body:    reqBody,
+		Headers: headers,
+	}, nil
+}
+
+// get ImpAlgoriXExt From First Imp. Only check and get first Imp.Ext.Bidder to ExtImpAlgorix
+func getImpAlgoriXExt(imp *openrtb2.Imp) (*openrtb_ext.ExtImpAlgorix, error) {
+	var extImpAlgoriX openrtb_ext.ExtImpAlgorix
+	var extBidder adapters.ExtImpBidder
+	err := json.Unmarshal(imp.Ext, &extBidder)
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal(extBidder.Bidder, &extImpAlgoriX)
+	if err != nil {
+		return nil, err
+	}
+	return &extImpAlgoriX, nil
+}
+
+func (a *adapter) getEndPoint(ext *openrtb_ext.ExtImpAlgorix) (string, error) {
+	endPointParams := macros.EndpointTemplateParams{SourceId: ext.Sid, AccountID: ext.Token}
+	return macros.ResolveMacros(a.EndpointTemplate, endPointParams)
+}
+
+func preProcess(request *openrtb2.BidRequest) {
+	for i := range request.Imp {
+		if request.Imp[i].Banner != nil {
+			banner := *request.Imp[i].Banner
+			if (banner.W == nil || banner.H == nil || *banner.W == 0 || *banner.H == 0) && len(banner.Format) > 0 {
+				firstFormat := banner.Format[0]
+				banner.W = &firstFormat.W
+				banner.H = &firstFormat.H
+				request.Imp[i].Banner = &banner
+			}
+		}
+	}
+}
+
+func (a *adapter) MakeBids(internalRequest *openrtb2.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: %d.", response.StatusCode),
+		}}
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d.", response.StatusCode),
+		}}
+	}
+
+	var bidResp openrtb2.BidResponse
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(1)
+
+	for _, seatBid := range bidResp.SeatBid {
+		for idx := range seatBid.Bid {
+			mediaType := getBidType(seatBid.Bid[idx].ImpID, internalRequest.Imp)
+			bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+				Bid:     &seatBid.Bid[idx],
+				BidType: mediaType,
+			})
+		}
+	}
+	return bidResponse, nil
+}
+
+func getBidType(impId string, imps []openrtb2.Imp) openrtb_ext.BidType {
+	mediaType := openrtb_ext.BidTypeBanner
+	for _, imp := range imps {
+		if imp.ID == impId {
+			if imp.Banner != nil {
+				return openrtb_ext.BidTypeBanner
+			}
+			if imp.Native != nil {
+				return openrtb_ext.BidTypeNative
+			}
+			if imp.Video != nil {
+				return openrtb_ext.BidTypeVideo
+			}
+		}
+	}
+	return mediaType
```

Comment:
We could remove `mediaType`. Unless you believe it makes for a more straightforward read, I'm good either way.
```
147   func getBidType(impId string, imps []openrtb2.Imp) openrtb_ext.BidType {
148 -     mediaType := openrtb_ext.BidTypeBanner
149       for _, imp := range imps {
150           if imp.ID == impId {
151               if imp.Banner != nil {
152 -                 return openrtb_ext.BidTypeBanner
    +                 break
153               }
154               if imp.Native != nil {
155                   return openrtb_ext.BidTypeNative
156               }
157               if imp.Video != nil {
158                   return openrtb_ext.BidTypeVideo
159               }
160           }
161       }
162 -     return mediaType
    +     return openrtb_ext.BidTypeBanner
163   }
adapters/algorix/algorix.go
```